File: Sources/PDFjet/Box.swift
/**
 * Box.swift
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */

/**
 * Used to create rectangular boxes on a page.
 */
public class Box : Drawable {
    var x: Float = 0.0
    var y: Float = 0.0

    var uri: String?
    var key: String?

    private var w: Float = 0.0
    private var h: Float = 0.0
    private var r: Float = 0.0

    private var color = Color.black

    private var width: Float = 0.0
    private var pattern: String = "[] 0"
    private var fillShape: Bool = false

    private var language: String?
    private var actualText: String = Single.space
    private var altDescription: String = Single.space

    /// Creates a box.
    public init() {
    }

    /**
     * Creates a box object.
     *
     * - Parameter x: the x coordinate of the top left corner of this box when drawn on the page.
     * - Parameter y: the y coordinate of the top left corner of this box when drawn on the page.
     * - Parameter w: the width of this box.
     * - Parameter h: the height of this box.
     */
    public init(
            _ x: Float,
            _ y: Float,
            _ w: Float,
            _ h: Float) {
        self.x = x
        self.y = y
        self.w = w
        self.h = h
    }

    /**
     * Sets the location of this box on the page.
     *
     * - Parameter x: the x coordinate of the top left corner of this box when drawn on the page.
     * - Parameter y: the y coordinate of the top left corner of this box when drawn on the page.
     */
    @discardableResult
    public func setLocation(
            _ x: Float,
            _ y: Float) -> Self {
        self.x = x
        self.y = y
        return self
    }

    /**
     * Sets the size of this box.
     *
     * - Parameter w: the width of this box.
     * - Parameter h: the height of this box.
     */
    @discardableResult
    public func setSize(
            _ w: Float,
            _ h: Float) -> Box {
        self.w = w
        self.h = h
        return self
    }

    /**
     * Sets the color for this box.
     *
     * - Parameter color: the color specified as an integer.
     */
    @discardableResult
    public func setColor(_ color: Int32) -> Box {
        self.color = color
        return self
    }

    /**
     * Sets the width of this line.
     *
     * - Parameter width: the width.
     */
    @discardableResult
    public func setLineWidth(_ width: Float) -> Box {
        self.width = width
        return self
    }

    /**
     * Sets the corner radius of this box.
     *
     * - Parameter r: the radius of the rounded corners.
     */
    @discardableResult
    public func setCornerRadius(_ r: Float) -> Box {
        self.r = r
        return self
    }

    /**
     * Sets the URI for the "click box" action.
     *
     * - Parameter uri: the URI
     */
    @discardableResult
    public func setURIAction(_ uri: String) -> Box {
        self.uri = uri
        return self
    }

    /**
     * Sets the destination key for the action.
     *
     * - Parameter key: the destination name.
     */
    @discardableResult
    public func setGoToAction(_ key: String) -> Box {
        self.key = key
        return self
    }

    /**
     * Sets the alternate description of this box.
     *
     * - Parameter altDescription: the alternate description of the box.
     * - Returns: this Box.
     */
    @discardableResult
    public func setAltDescription(
            _ altDescription: String) -> Box {
        self.altDescription = altDescription
        return self
    }

    /**
     * Sets the actual text for this box.
     *
     * - Parameter actualText: the actual text for the box.
     * - Returns: this Box.
     */
    @discardableResult
    public func setActualText(
            _ actualText: String) -> Box {
        self.actualText = actualText
        return self
    }

    /**
     * The line dash pattern controls the pattern of dashes and gaps used to stroke paths.
     * It is specified by a dash array and a dash phase.
     * The elements of the dash array are positive numbers that specify the lengths of
     * alternating dashes and gaps.
     * The dash phase specifies the distance into the dash pattern at which to start the dash.
     * The elements of both the dash array and the dash phase are expressed in user space units.
     * Examples of line dash patterns:
     *
     * ```
     *     "[Array] Phase"     Appearance          Description
     *     _______________     _________________   ____________________________________
     *
     *     "[] 0"              -----------------   Solid line
     *     "[3] 0"             ---   ---   ---     3 units on, 3 units off, ...
     *     "[2] 1"             -  --  --  --  --   1 on, 2 off, 2 on, 2 off, ...
     *     "[2 1] 0"           -- -- -- -- -- --   2 on, 1 off, 2 on, 1 off, ...
     *     "[3 5] 6"             ---     ---       2 off, 3 on, 5 off, 3 on, 5 off, ...
     *     "[2 3] 11"          -   --   --   --    1 on, 3 off, 2 on, 3 off, 2 on, ...
     * ```
     *
     * - Parameter pattern: the line dash pattern.
     */
    @discardableResult
    public func setPattern(
            _ pattern: String) -> Box {
        self.pattern = pattern
        return self
    }

    /**
     * Sets the private fillShape variable.
     * If the value of fillShape is true - the box is filled with the current brush color.
     *
     * - Parameter fillShape: the value used to set the private fillShape variable.
     */
    @discardableResult
    public func setFillShape(
            _ fillShape: Bool) -> Box {
        self.fillShape = fillShape
        return self
    }

    /**
     * Scales this box by the specified factor.
     *
     * - Parameter factor: the factor used to scale the box.
     */
    @discardableResult
    public func scaleBy(_ factor: Float) -> Box {
        self.x = self.x * factor
        self.y = self.y * factor
        return self
    }

    /**
     * Draws this box on the specified page.
     *
     * - Parameter page: the page to draw this box on.
     * - Returns: x and y coordinates of the bottom right corner of this component.
     */
    @discardableResult
    public func drawOn(_ page: Page?) -> [Float] {
        page!.addBMC(StructElem.P, language, actualText, altDescription)
        page!.setPenWidth(width)
        page!.setStrokeDashPattern(pattern)
        if fillShape {
            page!.setBrushColor(color)
        } else {
            page!.setPenColor(color)
        }
        if r == 0.0 {
            page!.moveTo(x, y)
            page!.lineTo(x + w, y)
            page!.lineTo(x + w, y + h)
            page!.lineTo(x, y + h)
            if fillShape {
                page!.fillPath()
            } else {
                page!.closePath()
            }
        } else {
            let k: Float = 0.55228
            var points: [Point] = []
            points.append(Point(x + r, y))
            points.append(Point((x + w) - r, y))
            points.append(Point((x + w - r) + r * k, y, Point.controlPointC))
            points.append(Point(x + w, (y + r) - r * k, Point.controlPointC))
            points.append(Point(x + w, y + r))
            points.append(Point(x + w, (y + h) - r))
            points.append(Point(x + w, ((y + h) - r) + r * k, Point.controlPointC))
            points.append(Point(((x + w) - r) + r * k, y + h, Point.controlPointC))
            points.append(Point((x + w) - r, y + h))
            points.append(Point(x + r, y + h))
            points.append(Point((x + r) - r * k, y + h, Point.controlPointC))
            points.append(Point(x, ((y + h) - r) + r * k, Point.controlPointC))
            points.append(Point(x, (y + h) - r))
            points.append(Point(x, y + r))
            points.append(Point(x, (y + r) - r * k, Point.controlPointC))
            points.append(Point((x + r) - r * k, y, Point.controlPointC))
            points.append(Point(x + r, y))
            page!.drawPath(points, fillShape ? PathOperator.fill : PathOperator.stroke)
        }
        page!.addEMC()

        if uri != nil || key != nil {
            page!.addAnnotation(Annotation(
                    Annotation.Link,
                    x,
                    y,
                    x + w,
                    y + h,
                    nil,    // Vertices
                    nil,    // Fill Color
                    0.0,    // Transparency
                    nil,    // Title
                    nil,    // Contents
                    uri,
                    key,    // The destination name
                    language,
                    actualText,
                    altDescription))
        }

        return [x + w, y + h]
    }
}   // End of Box.swift

File: net/pdfjet/Box.cs
/*
 * Box.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;

namespace PDFjet.NET {
/// <summary>
/// Used to create rectangular boxes on a page.
/// </summary>
public class Box : IDrawable {
    internal float x;
    internal float y;

    private float w;
    private float h;

    private int color = Color.black;
    private float width = 0f;
    private String pattern = "[] 0";
    private bool fillShape = false;

    private String uri = null;
    private String key = null;
    private String language = null;
    private String actualText = Single.space;
    private String altDescription = Single.space;

    /// <summary>
    /// The default constructor.
    /// </summary>
    public Box() {
    }

    /*
     * Creates a box object.
     *
     * @param x the x coordinate of the top left corner of this box when drawn on the page.
     * @param y the y coordinate of the top left corner of this box when drawn on the page.
     * @param w the width of this box.
     * @param h the height of this box.
     */
//    public Box(double x, double y, double w, double h) : this((float) x, (float) y, (float) w, (float) h) {
//    }

    /// <summary>
    /// Creates a box object.
    /// </summary>
    /// <param name="x">the x coordinate of the top left corner of this box when drawn on the page.</param>
    /// <param name="y">the y coordinate of the top left corner of this box when drawn on the page.</param>
    /// <param name="w">the width of this box.</param>
    /// <param name="h">the height of this box.</param>
    public Box(float x, float y, float w, float h) {
        this.x = x;
        this.y = y;
        this.w = w;
        this.h = h;
    }

    IDrawable IDrawable.SetLocation(float x, float y) {
        return SetLocation(x, y);
    }

    /// <summary>Sets the location of this box. Same as SetLocation.</summary>
    public Box SetXY(float x, float y) {
        SetLocation(x, y);
        return this;
    }

    /// <summary>
    /// Sets the location of this box on the page.
    /// </summary>
    /// <param name="x">the x coordinate of the top left corner of this box when drawn on the page.</param>
    /// <param name="y">the y coordinate of the top left corner of this box when drawn on the page.</param>
    public Box SetLocation(float x, float y) {
        this.x = x;
        this.y = y;
        return this;
    }

    /// <summary>Sets the location of the top left corner of this box.</summary>
    public Box SetLocation(double x, double y) {
        return SetLocation((float) x, (float) y);
    }

    /// <summary>
    /// Sets the size of this box.
    /// </summary>
    /// <param name="w">the width of this box.</param>
    /// <param name="h">the height of this box.</param>
    /// <returns>this Box object.</returns>
    public Box SetSize(double w, double h) {
        SetSize((float) w, (float) h);
        return this;
    }

    /// <summary>
    /// Sets the size of this box.
    /// </summary>
    /// <param name="w">the width of this box.</param>
    /// <param name="h">the height of this box.</param>
    /// <returns>this Box object.</returns>
    public Box SetSize(float w, float h) {
        this.w = w;
        this.h = h;
        return this;
    }

    /// <summary>
    /// Sets the color for this box.
    /// </summary>
    /// <param name="color">the color specified as an integer.</param>
    /// <returns>this Box object.</returns>
    public Box SetColor(int color) {
        this.color = color;
        return this;
    }

    /// <summary>
    /// Sets the width of this line.
    /// </summary>
    /// <param name="width">the width.</param>
    /// <returns>this Box object.</returns>
    public Box SetLineWidth(double width) {
        this.width = (float) width;
        return this;
    }

    /// <summary>
    /// Sets the width of this line.
    /// </summary>
    /// <param name="width">the width.</param>
    /// <returns>this Box object.</returns>
    public Box SetLineWidth(float width) {
        this.width = width;
        return this;
    }

    /// <summary>
    /// Sets the URI for the "click box" action.
    /// </summary>
    /// <param name="uri">the URI</param>
    /// <returns>this Box object.</returns>
    public Box SetURIAction(String uri) {
        this.uri = uri;
        return this;
    }

    /// <summary>
    /// Sets the destination key for the action.
    /// </summary>
    /// <param name="key">the destination name.</param>
    /// <returns>this Box object.</returns>
    public Box SetGoToAction(String key) {
        this.key = key;
        return this;
    }

    /// <summary>
    /// Sets the alternate description of this box.
    /// </summary>
    /// <param name="altDescription">the alternate description of the box.</param>
    /// <returns>this Box.</returns>
    public Box SetAltDescription(String altDescription) {
        this.altDescription = altDescription;
        return this;
    }

    /// <summary>
    /// Sets the actual text for this box.
    /// </summary>
    /// <param name="actualText">the actual text for the box.</param>
    /// <returns>this Box.</returns>
    public Box SetActualText(String actualText) {
        this.actualText = actualText;
        return this;
    }

    /// <summary>
    /// The line dash pattern controls the pattern of dashes and gaps used to stroke paths.
    /// It is specified by a dash array and a dash phase.
    /// The elements of the dash array are positive numbers that specify the lengths of
    /// alternating dashes and gaps.
    /// The dash phase specifies the distance into the dash pattern at which to start the dash.
    /// The elements of both the dash array and the dash phase are expressed in user space units.
    /// <code>
    /// Examples of line dash patterns:
    ///
    ///     "[Array] Phase"     Appearance          Description
    ///     _______________     _________________   ____________________________________
    ///
    ///     "[] 0"              -----------------   Solid line
    ///     "[3] 0"             ---   ---   ---     3 units on, 3 units off, ...
    ///     "[2] 1"             -  --  --  --  --   1 on, 2 off, 2 on, 2 off, ...
    ///     "[2 1] 0"           -- -- -- -- -- --   2 on, 1 off, 2 on, 1 off, ...
    ///     "[3 5] 6"             ---     ---       2 off, 3 on, 5 off, 3 on, 5 off, ...
    ///     "[2 3] 11"          -   --   --   --    1 on, 3 off, 2 on, 3 off, 2 on, ...
    /// </code>
    /// </summary>
    /// <param name="pattern">the line dash pattern.</param>
    /// <returns>this Box object.</returns>
    public Box SetPattern(String pattern) {
        this.pattern = pattern;
        return this;
    }

    /// <summary>
    /// Sets the private fillShape variable.
    /// If the value of fillShape is true - the box is filled with the current brush color.
    /// </summary>
    /// <param name="fillShape">the value used to set the private fillShape variable.</param>
    /// <returns>this Box object.</returns>
    public Box SetFillShape(bool fillShape) {
        this.fillShape = fillShape;
        return this;
    }

    /// <summary>
    /// Scales this box by the specified factor.
    /// </summary>
    /// <param name="factor">the factor used to scale the box.</param>
    public void ScaleBy(double factor) {
        ScaleBy((float) factor);
    }

    /// <summary>
    /// Scales this box by the specified factor.
    /// </summary>
    /// <param name="factor">the factor used to scale the box.</param>
    public void ScaleBy(float factor) {
        this.x *= factor;
        this.y *= factor;
    }

    /// <summary>
    /// Draws this box on the specified page.
    /// </summary>
    /// <param name="page">the page to draw on.</param>
    /// <returns>x and y coordinates of the bottom right corner of this component.</returns>
    public float[] DrawOn(Page page) {
        page.AddBMC(StructElem.P, language, actualText, altDescription);
        page.SetPenWidth(width);
        page.SetStrokeDashPattern(pattern);
        if (fillShape) {
            page.SetBrushColor(color);
        } else {
            page.SetPenColor(color);
        }
        page.MoveTo(x, y);
        page.LineTo(x + w, y);
        page.LineTo(x + w, y + h);
        page.LineTo(x, y + h);
        if (fillShape) {
            page.FillPath();
        } else {
            page.ClosePath();
        }
        page.AddEMC();

        if (uri != null || key != null) {
            page.AddAnnotation(new Annotation(
                    Annotation.Link,
                    x,
                    y,
                    x + w,
                    y + h,
                    null,   // Vertices
                    null,   // Fill Color
                    0f,     // Transparency
                    null,   // Title
                    null,   // Contents
                    uri,
                    key,    // The destination name
                    language,
                    actualText,
                    altDescription));
        }

        return new float[] {x + w, y + h + width};
    }
}   // End of Box.cs
}   // End of namespace PDFjet.NET
